storybookjs/storybook
# Core: Allow a teardown function to be returned from `renderToDOM`

Issue: Stories (both in Canvas and when rendered in `modernInlineRender`) had no way to be properly torn down, leading to memory/event listener leaks when browsing around (as noticed in Angular by @kroeder).

## What I did

Added some stories to highlight the problem to `react-ts` (not sure this is where they should go, and if we should have a way to properly test this @shilman).

Notice the following (without the changes to `react/../render.tsx`):

1. Browse to a buttonX story
2. Go to docs tab
3. Notice event listener for Canvas is still active (you will see click handlers fire one more time than the number of rendered stories in the docs page).
4. Go to docs tab for other buttonX story
5. Notice event listeners from previous docs page are not torn down.

The Canvas event listeners eventually do get torn down because we are ultimately rendering a different story to the same DOM element, and React handles this for us.

But
a) it should happen earlier I think and
b) that doesn't happen automatically in Angular as I understand it.

## Solution

I allowed `renderToDOM` to return a `teardown` function that will be called whenever a `StoryRender` is taken off the screen.

## How to test


See reproduction above. I will also add a bunch of `PreviewWeb` integration tests once we are satisfied this is the right approach.
